### PR TITLE
Refactoring collect token sorts and attributes

### PR DESCRIPTION
**Pull request by Robertorosmaninho**
Part of #3444

This PR is part of the effort to refactor the [`ModuleToKore.java`](https://github.com/runtimeverification/k/blob/master/kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java) to be simpler and only contains the AST traversing to print out the kore syntax. The core of this modification is to create a new compiler pass for pre-compute the `SortTokens` and `Attributes` on the main module before the call of `ModuleToKore`.

### CODE REVIEW

File: kore/src/main/scala/org/kframework/definition/outer.scala [hunk 1]
```diff
@@ -309,6 +308,45 @@ case class Module(val name: String, val imports: Set[Import], localSentences: Se
     }
     functionRules
   }
+
+
+  val hasAttributesMap: mutable.Map[Att.Key, java.lang.Boolean] = mutable.Map[Att.Key, java.lang.Boolean]()
```
Comment (radumereuta) [inline kore/src/main/scala/org/kframework/definition/outer.scala]: One of the benefits of using scala is that it ensures immutability.
I would strongly avoid using mutable classes here.

Comment (dwightguth) [reply kore/src/main/scala/org/kframework/definition/outer.scala]: Agree 100%. We should be using the existing attributes, or else a lazy val that computes some information about the module.

**Review by dwightguth — CHANGES_REQUESTED** [review]
See my comment

File: kore/src/main/scala/org/kframework/definition/outer.scala (hunk 1 above)
Comment (Robertorosmaninho) [reply kore/src/main/scala/org/kframework/definition/outer.scala]: The problem is that we depend on the result of `RuleInfo` and `ExpandMacros` to know whether we should have an attribute on that list. Also, we depend on the `heatCoolEq` value.

Comment (Robertorosmaninho) [reply kore/src/main/scala/org/kframework/definition/outer.scala]: These informations comes outside the information available in the module itself.

Comment (radumereuta) [reply kore/src/main/scala/org/kframework/definition/outer.scala]: If this information is required in only one step, you can compute it locally.
Pretty much what RuleInfo does already. (IIRC)

**Comment by Baltoli** [comment]
Superseded by more recent work on ModuleToKore.

**Pull request closed without merging**